File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx = [0,0,-1,1]
dy = [-1,1,0,0]

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    que = deque()
    que.append((0,0,1))
    while que:
        x,y,cnt = que.popleft()
        if x == n-1 and y == m-1:
            return cnt
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<=nx<n and 0<=ny<m: # 범위 초과하지  않고
                if maps[nx][ny] == 1: # 아직 방문하지 않았다면
                    maps[nx][ny] = cnt
                    que.append((nx,ny,cnt+1))
                    maps[x][y] = 0 # 현재 위치 벽으로 만들어줘야 역행 안함
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.io.*;
import java.math.*;
import java.util.*;

class Solution { 

    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};


    public int solution(int[][] maps) {
        ArrayDeque<int[]> que = new ArrayDeque<>();
        int n = maps.length;
        int m = maps[0].length;
        
        que.offer(new int[]{0,0,1});

        while (!que.isEmpty()) {
            int[] arr = que.poll();
            int x = arr[0];
            int y = arr[1];
            int cnt = arr[2];

            if (x==n-1 && y==m-1){
                return cnt;
            }

            for (int i=0;i<4;i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (0<=nx && nx<n && 0<=ny && ny<m){
                    if (maps[nx][ny]==1){
                        maps[nx][ny] = cnt;
                        que.offer(new int[]{nx,ny,cnt+1});
                        maps[nx][ny]=0;
                    }
                }

            }

        }
        return -1;
    }
}